### DISCUSSION TITLE
PatchDataset/GridPatchDataset or SlidingWindowInferer for patch-based train with large volume

### DISCUSSION THREAD

**Discussion by Littlehhh** [Project-MONAI/MONAI#1484] [Q&A]
# Description
Hi
I found that there are several ways in MONAI to implement patch-based(grid patch) training for large medical volume.
1. `PatchDataset/GridPatchDataset` with `SimpleInferer`
2. `Dataset` with `SlidingWindowInferer`
3. `Dataset` with `RandCropByPosNegLabel/(or extra transfrom)`

So which way should be better? And these ways have different loss(single patch/full volume/several patch) backpropagations.
I am confused about this and I think maybe the first way is a more reasonable way for patch-based system.

In addition, from the perspective of framework design, which way is more appropriate, and what are the uses of other ways for patch-based train?

Thanks.

**Comment by Nic-Ma** [comment]
Hi @Littlehhh ,

Thanks for your interest here.
1. `RandCropByPosNegLabel` is a transform and usually used to generate random patches during training, the following transforms will apply to every patch separately.
2. `SlidingWindowInfer` is used to split big image into patches for the network forward(), so it doesn't contain random logic and the input data is already handled by transforms, usually used for validation.
3. `PatchDataset` can do global shuffling on the patches.

You can start from this example:
https://github.com/Project-MONAI/tutorials/blob/master/3d_segmentation/spleen_segmentation_3d.ipynb

Thanks.

**Reply by Littlehhh** [reply]
Hi @Nic-Ma 
Thanks for your reply.

Both `PatchDataset` and  `SlidingWindowInferer` are  for large volume patch-based process. 
Is it possible to integrated  these two functions  together ？
`PatchDataset` has already prepared the patch so `SlidingWindowInferer` is a little bit redundant and  it introduces one more module to the MONAI. Is this module necessary for MONAI.

Thanks in advance.

**Comment by Nic-Ma** [comment]
Hi @Littlehhh ,

`SlidingWindow` is a very common inference method, especially for medical images, I think it's necessary to provide this feature.

Thanks.

**Reply by Littlehhh** [reply]
Hi @Nic-Ma
Thanks for your reply.
Very happy to discuss this with you. 
I agree that `SlidingWindow` is a important function for medical images.
Is there any way to integrate these two component together？
Just like check if `PatchDataset` than use `slidingwindow`.
but  `PatchDataset` has already cropped the patch, and  `slidingwindow` do the same crop.
I think maybe there is a suitable way for Patch-based train and Inference. 

Thanks in advance.

**Comment by Nic-Ma** [comment]
Hi @wyli ,

Could you please help share more use cases about `PatchDataset`?

Thanks in advance.
